src/FGS.Extensions.Hosting.Middleware/MiddlewareApplyingHostDecorator.cs

```diff
@@ -0,0 +1,39 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+using FGS.Extensions.Hosting.Middleware.Abstractions;
+
+using Microsoft.Extensions.Hosting;
+
+namespace FGS.Extensions.Hosting.Middleware
+{
+    /// <summary>
+    /// An implementation of <see cref="IHost"/> that decorates an underlying instance with interceptor behavior represented
+    /// by a given instance of <see cref="IHostingMiddleware"/>.
+    /// </summary>
+    internal sealed class MiddlewareApplyingHostDecorator : IHost
+    {
+        private readonly IHost _decorated;
+        private readonly IHostingMiddleware _hostingMiddleware;
+
+        internal MiddlewareApplyingHostDecorator(IHost decorated, IHostingMiddleware hostingMiddleware)
+        {
+            _decorated = decorated;
+            _hostingMiddleware = hostingMiddleware;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
```

Comment:
Just so I'm sure, can you reaffirm that we don't need to dispose `_decorated` here because it was created by a service provider, which will be responsible for cleanup?

Reply:
Actually, instead of relying on an understanding of what `IServiceProvider` implementations might do, this class instead relies on the convention that you don't typically dispose something that you received as a constructor parameter. (*Except* for cases when an additional parameter is provided to toggle whether or not the constructor caller is indicating this new object will "own" the disposable thing. Not a common case in LOB code.)

However, your question prompted me to track down how the `IServiceProvider` is likely to behave, and I believe that yes, things will get disposed in the way that we are typically used to. Originally, disposal was an Autofac-specific feature that not all DI-containers shared. However, when Microsoft refactored their DI abstractions as part of making .NET Core, they based it (roughly) on Autofac and so it [seems to now be the case](https://stackoverflow.com/a/40845639) to expect `IServiceProvider` to _eventually_ clean up things it creates (at the end of the current scope.)

Reply:
:+1:

> this class instead relies on the convention that you don't typically dispose something that you received as a constructor parameter.

Didn't know about that, thanks for the insight!